app/models/linked_case.rb

```diff
@@ -0,0 +1,74 @@
+class LinkedCase < ApplicationRecord
+  self.inheritance_column = :_unused
+
+  attr_accessor :linked_case_number
+
+  belongs_to :case, class_name: 'Case::Base',
+             inverse_of: :linked_cases
+
+  belongs_to :linked_case, class_name: 'Case::Base'
+
+  enum type: {
+         related: 'related',
+         original: 'original',
+       }
+  before_validation :find_linked_case_by_number
+  after_create :create_reverse_link
+  after_destroy :destroy_reverse_link
+
+  validates :linked_case, presence: true
+  validates :linked_case_number,
+            presence: true,
+            if: -> { linked_case_id.nil? }
+  validates :linked_case_id,
+            presence: true,
+            if: -> { linked_case_number.nil? }
+
+  validates_with ::CaseLinkTypeValidator,
+                 if: -> { self.case_id.present? &&
+                          (self.linked_case_id.present? ||
+                           self.linked_case_number.present?) }
+  validate :validate_case_not_linked_back_to_itself,
+           if: -> { self.case_id.present? }
+
+  private
+
+  def find_linked_case_by_number
+    if linked_case_number.present?
+      self.linked_case = Case::Base.find_by(number: linked_case_number)
+      if self.linked_case.nil?
+        self.errors.add(:linked_case_number, :missing)
+      end
+    end
+  end
+
+  def validate_case_not_linked_back_to_itself
+    if linked_case_number.present?
+      if self.case.number == linked_case_number
+        errors.add(:linked_case_number, :references_self)
+      end
+    else
+      if self.case_id == linked_case_id
+        errors.add(:linked_case, :references_self)
+      end
+    end
+  end
+
+  def create_reverse_link
+    self.class.skip_callback(:create, :after, :create_reverse_link)
+    self.class.create(case_id: linked_case_id, linked_case_id: self.case_id)
+  rescue ActiveRecord::RecordNotUnique
+    nil
+  ensure
+    self.class.set_callback(:create, :after, :create_reverse_link)
+  end
```

Comment:
Wouldn't it be easier to see if the link was already there, and if not, create it?  That would save having to skip the callback, rescue RecordNotUnique and reset the call back.

Reply:
Hrm. I think I was trying to optimise to reduce DB access. Just to try it out, the code would look like:

```
def create_reverse_link
  unless self.class.exist(case_id: linked_case_id, linked_case_id: self.case_id)
    self.class.skip_callback(:create, :after, :create_reverse_link)
    self.class.create(case_id: linked_case_id, linked_case_id: self.case_id)
    self.class.set_callback(:create, :after, :create_reverse_link)
  end
end
```

A little bit tidier, but it isn't atomic and could cause introduce a race condition ... it should probably be protected with a rescue:

```
def create_reverse_link
  unless self.class.exist(case_id: linked_case_id, linked_case_id: self.case_id)
    begin
      self.class.skip_callback(:create, :after, :create_reverse_link)
      self.class.create(case_id: linked_case_id, linked_case_id: self.case_id)
    rescue ActiveRecord::RecordNotUnique
      nil
    ensure
      self.class.set_callback(:create, :after, :create_reverse_link)
    end
  end
end
```

What do you think?

Reply:
I was thinking more that the LinkCase Service should call something like `LinkedCases.link(case_a, case_b, link_type)`
 and that the LinkedCases.link method should take care of making both links in a transaction.  No need for after create hooks, which sometimes are more trouble than they are worth.